Felix Upload cog for uploading attached files to EMKC

dhdavvie opened the pull request:
Occasionally users submit files to the Discord, and some people may not want to download random files onto their system or simply prefer to have the code on a snippet site. This cog takes a message ID, looks for attachments, tries to match them to the language based on file extension and uploads the contents of the file to EMKC Snippets. 

There is a limit of 10MB for the size of the file, this seemed like more than enough for pretty much any file and helps reduce risk of overloading felix memory. Standard Discord upload limit is 8MB anyways, but Nitro gets 50MB. 

Could add the possibility to specify the language if we wanted to. Maybe for a future PR?

brtwrst commented:
1. Small thing first.
Consider squashing all PR commits into one and force pushing your branch. No reason to add 3 one line commits in a PR.

2. Big thing:
I went through the process of removing requests from the source code back in March.
Requests is not an async library (also urllib.request is not) and will halt the bot execution while it waits for responses.
Please remove the non-async parts of the new cog and replace them with the system in place (aiohttp). You can find numerous examples throughout the source code (git command, graph command ...) You can always ask here or on discord if you need help on that.

3. Another thing
I always thought a command like this should give everybody the ability to create emkc snippets from discord. Now admin interaction is required - which sucks. I think a command that everybody can use and that creates a snippet on emkc from message-text or an attached file would be far more useful. That way it could also be associated with a user. The way it is now, the Snippet is kind of orphaned and the only link in existence is the one that is returned from the command. I'm not sure if the snippet api supports these things yet (making felix create a snippet for a user) but that would make the command a lot more useful.

Maybe @realtux could clarify what the api can do and what it can't.

solidassassin commented:
We made `aiohttp.ClientSession` available from `client` which is reachable in every cog.

dhdavvie commented:
> Consider squashing all PR commits into one and force pushing your branch. No reason to add 3 one line commits in a PR.

Sure, I usually do it via "Squash and Merge" but happy to squash them on my end.

> We made aiohttp.ClientSession available from client which is reachable in every cog.

Amazing, will use that instead. Didn't realise the issues with request and async.

> I always thought a command like this should give everybody the ability to create emkc snippets from discord. Now admin interaction is required - which sucks.

This can be changed, I added it as admin only to begin with with the thought of not having this be abused by users, but its easy to just remove the admin check. Should we make it open to everyone?

> That way it could also be associated with a user. The way it is now, the Snippet is kind of orphaned and the only link in existence is the one that is returned from the command.

I agree that it would be incredibly useful to link the snippet to the Discord user assuming they are an EMKC member with their discord account linked. This could be future work, as we may also need to update EMKC so that we have an API to create snippets from a user.

brtwrst commented:
Well everyone can currently go to emkc and "abuse" snippets already so yeah  make it public.

I don't like merge commits i rebase PR's and sadly there is no squash and rebase option. ;)

brtwrst commented:
But when it's available to everyone it should be rewritten to accept code-block text or an attachment directly. Makes no sense to upload a file and then type the command to put it on emkc 1 message later. This should all happen with 1 message. This would also save you the step of `download_file_contents` because the file will already be in memory while processing the message.

dhdavvie commented:
> But when it's available to everyone it should be rewritten to accept code-block text or an attachment directly. Makes no sense to upload a file and then type the command to put it on emkc 1 message later. This should all happen with 1 message. This would also save you the step of `download_file_contents` because the file will already be in memory while processing the message.

The original thought was when another user uploads a file (as they sometimes do in order to share code) and you want to have that file be turned into a snippet without needing to download the file. For example if you're on mobile, with this you could upload someone's file without needing to ask them to do it. 

I don't quite see the need of uploading a text message to snippet, since thats already in a format that is readable without needing to download to computer, which is the number one problem I am trying to solve with this command.

brtwrst commented:
ok then i would suggest the following
1. i would rename the command to felix snippet (upload is too broad)
2. if the command was called without an attachment 
  -> assume it was called with the message id and proceed like you do now.
If it was called with an attachment 
  -> assume (maybe check) the attachment is a source code text-file and upload it directly

dhdavvie commented:
> i would rename the command to felix snippet (upload is too broad)

sounds good

> if the command was called without an attachment
-> assume it was called with the message id and proceed like you do now.
If it was called with an attachment
-> assume (maybe check) the attachment is a source code text-file and upload it directly

I think this is a good set of functionality, I will look into implementing this

dhdavvie commented:
Removed the admin check, added a brief, and made `message_id` optional. Omitting it will cause the bot to look at the current message for file attachments.

brtwrst commented:
nice one thanks
